sdk/web-pubsub/web-pubsub/src/hubClient.ts

```diff
@@ -0,0 +1,659 @@
+   * const client = new HubClient(endpoint, cred, 'chat');
+   * ```
+   *
+   * @param endpoint The endpoint to connect to
+   * @param credential An AzureKeyCredential holding your service key
+   * @param hubName The name of the hub to connect to.
+   * @param options Options to configure the http pipeline
+   */
+  constructor(
+    endpoint: string,
+    credential: AzureKeyCredential,
+    hubName: string,
+    options?: HubAdminClientOptions
+  );
+  constructor(
+    epOrCs: string,
+    credsOrHubName?: AzureKeyCredential | string,
+    hubNameOrOpts?: string | HubAdminClientOptions,
+    opts?: HubAdminClientOptions
+  ) {
+    // unpack constructor arguments
+    if (typeof credsOrHubName === "object" && "key" in credsOrHubName) {
+      this.endpoint = epOrCs;
+      this.credential = credsOrHubName;
+      this.hubName = hubNameOrOpts as string;
+      this.clientOptions = opts;
+    } else {
+      const parsedCs = parseConnectionString(epOrCs);
+      this.endpoint = parsedCs.endpoint;
+      this.credential = parsedCs.credential;
+      this.hubName = credsOrHubName as string;
+      this.clientOptions = hubNameOrOpts as PipelineOptions;
+    }
+
+    const internalPipelineOptions: InternalPipelineOptions = {
+      ...this.clientOptions,
+      ...{
+        loggingOptions: {
+          logger: logger.info
+        }
+      }
+    };
+
+    const pipeline = createPipelineFromOptions(
+      internalPipelineOptions,
+      webPubSubAzureKeyCredentialPolicyFactory(this.credential)
+    );
+
+    this.client = new GeneratedClient(this.endpoint, pipeline);
+  }
+
+  /**
+   * Auth the client connection with userId and custom claims if any
+   * @param options The options that the client has
+   */
+  public async getAuthenticationToken(
+    options?: GetAuthenticationTokenOptions
+  ): Promise<GetAuthenticationTokenResponse> {
+    const endpoint = this.endpoint.endsWith("/") ? this.endpoint : this.endpoint + "/";
+    const key = this.credential.key;
+    const hub = this.hubName;
+    var clientEndpoint = endpoint.replace(/(http)(s?:\/\/)/gi, "ws$2");
+    var clientUrl = `${clientEndpoint}client/hubs/${hub}`;
+    const audience = `${endpoint}client/hubs/${hub}`;
+    var payload = options?.claims ?? {};
+    var signOptions: jwt.SignOptions = {
+      audience: audience,
+      expiresIn: "1h",
+      algorithm: "HS256"
+    };
+    if (options?.userId) {
+      signOptions.subject = options?.userId;
+    }
+    const token = jwt.sign(payload, key, signOptions);
+    const url = `${clientUrl}?access_token=${token}`;
+    return {
+      baseUrl: clientUrl,
+      token: jwt.sign(payload, key, signOptions),
+      url: url
+    };
+  }
+
+  /**
+   * Get a client for a group
+   * @param groupName The name of the group to connect to.
+   */
+  public group(groupName: string): WebPubSubGroup {
+    const client = new WebPubSubGroup(this.hubName, groupName);
+    client["client"] = this.client;
```

Comment:
this feels strange, why not pass it into the constructor?

Reply:
I'll change this when I move to interface. At one point the class was public and could be constructed directly, and I was using this to sneak over the client so I wouldn't have to recreate it.